### ISSUE TITLE
Bug: preserve backslashes before non-ASCII punctuation in `no-trailing-heading-punctuation`

### ISSUE THREAD

**Issue by lumirlumir** [eslint-markdown/eslint-markdown#757]
### Environment

ESLint version: v9.39.5
@eslint/markdown version: v8.0.3
eslint-markdown version: v0.17.0
Node version: v24.18.0
npm version: v11.16.0
Operating System: Windows

### Which language are you using?

commonmark & gfm

### What did you do?

<details>
<summary>Configuration</summary>

```js
import { defineConfig } from 'eslint/config';
import markdown from '@eslint/markdown';
import md from './packages/eslint-markdown/build/index.js';

export default defineConfig([
  {
    files: ['**/*.md'],
    plugins: {
      markdown,
      md,
    },
    language: 'markdown/commonmark',
    rules: {
      'md/no-trailing-heading-punctuation': 'error',
    },
  },
]);
```
</details>

```md
# Hello\！
```

Run the following commands from the linked repository revision:

```sh
npm install
npm run build:pkg:em
npx eslint --fix repro.md
```

### What did you expect to happen?

The rule should remove the default full-width trailing punctuation but preserve the preceding literal backslash:

```md
# Hello\
```

CommonMark only recognizes backslash escapes before ASCII punctuation. Full-width and custom Unicode punctuation are not backslash-escapable.

### What actually happened?

The fixer removes both the full-width punctuation and the preceding literal backslash:

```md
# Hello
```

The fixer currently treats every odd run of adjacent backslashes as an escape without checking whether the first matched punctuation character is CommonMark-escapable ASCII punctuation.

### Link to Minimal Reproducible Example

N/A

### Participation

- [ ] I am willing to submit a pull request for this issue.

### AI acknowledgment

- [ ] No - I did not use AI to generate this issue report.
- [ ] Yes - minimal assistance (e.g., IDE autocomplete, small code completions, grammar fixes).
- [x] Yes - substantive assistance (issue drafting with Codex).

### Additional comments

The fix should only consume the immediately preceding escape backslash when the first matched punctuation character is CommonMark-escapable ASCII punctuation. Regression tests should cover default full-width punctuation, custom Unicode punctuation, and existing ASCII escaped punctuation such as `\!`.

**Comment by lumirlumir** [comment]
This is an extreme edge case, thus low priority.

**Comment by kevin-lozada-santos** [comment]
I can check this on current main and prepare a focused fix. I'll verify literal backslashes before default full-width and custom Unicode punctuation in both CommonMark and GFM, with controls for escaped ASCII punctuation and even/odd backslash runs, then return the reproduction and validation results.

**Comment by kevin-lozada-santos** [comment]
Confirmed on `c76c55cfa6a6fbbbcb356e00dae84d0451548a48` and submitted the fix in #760 (`b93c2348b3539b08d82753f169c9b2c086e7f238`).

The odd-backslash check was consuming a literal backslash before Unicode punctuation. The fix checks the first removed character against CommonMark's escapable ASCII punctuation set. It preserves `# Hello\！` as `# Hello\`, while existing ASCII cases such as `# Hello\!` still become `# Hello`.

Validation on Windows / Node 22.23.1:
- The added cases produced 12 expected failures before the fix across CommonMark and GFM; all 203 checks for this rule now pass.
- Full type checks and 2,873 tests pass; package/website builds and all repository lint checks pass.
- Twenty checks against the built plugin verify actual autofix output and unchanged second-pass output, including custom Unicode, mixed punctuation ordering, backslash parity, and closed/Setext headings.

I used [Brain Scanner](https://brainscanner.dev/) to organize the relevant source map and test evidence, then checked its graph and Reports view in the browser. The test results above come from executing ESLint and the repository's checks.

**Comment by khakaa** [comment]
Hi @lumirlumir, I'd like to work on this. Could you assign it to me?

**Comment by tooth-is-silver** [comment]
Thanks for your interest! I've assigned this issue to @khakaa.
A PR is welcome!

**Comment by lumirlumir** [comment]
TIP: Re-using one of the following constant would be helpful in this case:

https://github.com/eslint-markdown/eslint-markdown/blob/ad1be4f3910185a93f41696bf1afde95e99487fb/packages/eslint-markdown/src/core/constants.ts#L21-L42
